Source/WebCore/Modules/gamepad/GamepadHapticActuator.cpp

```diff
@@ -28,34 +28,75 @@
 #if ENABLE(GAMEPAD)
 #include "GamepadHapticActuator.h"
 
+#include "Document.h"
+#include "EventLoop.h"
+#include "Gamepad.h"
 #include "GamepadEffectParameters.h"
+#include "GamepadProvider.h"
 #include "JSDOMPromiseDeferred.h"
+#include <wtf/CompletionHandler.h>
 
 namespace WebCore {
 
-Ref<GamepadHapticActuator> GamepadHapticActuator::create()
+Ref<GamepadHapticActuator> GamepadHapticActuator::create(Gamepad& gamepad)
 {
-    return adoptRef(*new GamepadHapticActuator);
+    return adoptRef(*new GamepadHapticActuator(gamepad));
 }
 
-GamepadHapticActuator::GamepadHapticActuator()
+GamepadHapticActuator::GamepadHapticActuator(Gamepad& gamepad)
     : m_type { Type::Vibration }
+    , m_gamepad { gamepad }
 {
 }
 
-bool GamepadHapticActuator::canPlayEffectType(EffectType) const
+GamepadHapticActuator::~GamepadHapticActuator() = default;
+
+bool GamepadHapticActuator::canPlayEffectType(EffectType effectType) const
 {
-    return false;
+    return m_gamepad && m_gamepad->supportedEffectTypes().contains(effectType);
 }
 
-void GamepadHapticActuator::playEffect(EffectType, GamepadEffectParameters&&, Ref<DeferredPromise>&& promise)
+void GamepadHapticActuator::playEffect(Document& document, EffectType effectType, GamepadEffectParameters&& effectParameters, Ref<DeferredPromise>&& promise)
 {
-    promise->reject(Exception { NotSupportedError });
+    if (!document.isFullyActive() || document.hidden() || !m_gamepad) {
+        promise->resolve<IDLEnumeration<Result>>(Result::Preempted);
+        return;
+    }
+    if (auto playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr)) {
+        document.eventLoop().queueTask(TaskSource::Gamepad, [playingEffectPromise = WTFMove(playingEffectPromise)] {
+            playingEffectPromise->resolve<IDLEnumeration<Result>>(Result::Preempted);
+        });
+    }
+    if (!canPlayEffectType(effectType)) {
+        promise->reject(Exception { NotSupportedError, "This gamepad doesn't support playing such effect"_s });
+        return;
+    }
+    m_playingEffectPromise = WTFMove(promise);
+    GamepadProvider::singleton().playEffect(m_gamepad->index(), m_gamepad->id(), effectType, effectParameters, [this, protectedThis = Ref { *this }, document = Ref { document }, playingEffectPromise = m_playingEffectPromise](bool success) {
+        if (m_playingEffectPromise != playingEffectPromise)
+            return; // Was already pre-empted.
+        document->eventLoop().queueTask(TaskSource::Gamepad, [playingEffectPromise = std::exchange(m_playingEffectPromise, nullptr), success] {
```

Comment:
Shouldn't you do the same nullptr check for `playingEffectPromise` you use on line 65 here? Why wouldn't it be at the same risk?

Reply:
On the previous line, we just ensured that `m_playingEffectPromise == playingEffectPromise` so it is safe to assume `m_playingEffectPromise` is not null.